cmd/tempo/app/app.go

```diff
@@ -161,16 +168,18 @@ func newDefaultConfig() *Config {
 type App struct {
 	cfg Config
 
-	Server       *server.Server
-	ring         *ring.Ring
-	overrides    *overrides.Overrides
-	distributor  *distributor.Distributor
-	querier      *querier.Querier
-	frontend     *frontend_v1.Frontend
-	compactor    *compactor.Compactor
-	ingester     *ingester.Ingester
-	store        storage.Store
-	MemberlistKV *memberlist.KVInitService
+	Server        *server.Server
+	ring          *ring.Ring // TODO should we rename this to ingesterRing?
+	generatorRing *ring.Ring
```

Comment:
i believe that the ingester ring is shared here is b/c it's used by both distribs and queriers. this ring can probably remain local to the initGenerator() method.

Reply:
Well spotted! I'll remove it from the struct.

Reply:
Okay, I've tried some options here and it's not super clear cut what the best option is. `generatorRing` is actually only used by the distributor, the distributor uses it as a `ring.ReadRing`.

First issue: `generatorRing` is a service, by returning it in `initGeneratorRing` it becomes its own module which updates the service. If we create the ring within `initDistributor` we have to use a `services.Manager` to wrap both services. This makes stuff more complicated.

Second issue: we could move the ring creation into `distributor.New`. This should work, but we would have to pass the `metrics_generator.ring` config in there as well because we need this to create the ring. So this would create a dependency from the `distributor` to the `generator` package.

We could avoid this by creating a new top-level config block `metrics_generator_ring`. We already have `ingester_client` and `metrics_generator_client`.
I think this would make it more clear that this config is shared by multiple components (the distributor and the generator), but it's yet another top-level config block. Thoughts?